### PR TITLE
05 15 feat accounting event pipeline

### LINKED ISSUE #3357: [Bug]: Bifrost Billing Discrepancy Analysis Report
### Prerequisites

- [x] I have searched existing issues and discussions to avoid duplicates
- [x] I am using the latest version (or have tested against main/nightly)

## Background

Claude Code requests are proxied through Bifrost to Anthropic, using the model `anthropic/claude-opus-4-6`.

Over a 30-day window, the following billing discrepancy was observed:
- **Claude Code / Anthropic side**: ~$120
- **Bifrost Governance plugin statistics**: ~$90
- **Difference**: ~$30 (about 25%)

---

## Causes Ruled Out

| Investigation Item | Conclusion | Notes |
|--------------------|------------|-------|
| Pricing data error | **Ruled out** | The Opus 4.6 pricing in Bifrost's datasheet matches Anthropic's actual rates ($5/$25/MTok input/output, $0.50 cache read, $6.25 cache write) |
| Prompt caching calculation error | **Ruled out** | Cache read/write token decomposition and rate mapping in `computeTextCost()` are correct (`pricing.go:506`) |
| Responses API not billed | **Ruled out** | `/v1/responses` goes through the same `RunPostLLMHooks` pipeline; usage extraction is complete |
| Anthropic usage field mapping lost | **Ruled out** | `input_tokens`, `cache_read_input_tokens`, `cache_creation_input_tokens` are all parsed correctly (`anthropic/chat.go:886`) |
| Pricing resolution falling back to wrong model | **Ruled out** | `resolvePricing()` has no special fallback chain for the Anthropic provider; it precisely matches `claude-opus-4-6|anthropic|chat` |

---

## Root Cause: Difference in Accounting Scope

**Bifrost only counts the cost of "successful requests", but Anthropic charges for every request whose tokens were processed (including failed ones).**

The core code lives in two places:

### 1. The success determination

`plugins/governance/main.go:1588`:
```go
success := (result != nil)  // a nil result counts as failure
```

All error paths (timeout, cancellation, interruption, 500) only pass `err` and not `result`, which makes `success = false`.

### 2. Failed requests skip billing

`plugins/governance/tracker.go:72`:
```go
func (t *UsageTracker) UpdateUsage(ctx context.Context, update *UsageUpdate) {
    if !update.Success {
        t.logger.Debug("Request was not successful, skipping usage update")
        return  // ← all costs are lost
    }
    // ...
}
```

---

### Scenario 1: Client disconnect (streaming interruption)

- **Location**: `plugins/governance/main.go:1588`
- **Trigger**: Client disconnects in the middle of a streaming response (Ctrl+C, closing the terminal, etc.)
- **Mechanism**: The Anthropic provider's defer calls `HandleStreamCancellation()` (`providers/utils/utils.go:2139`), which propagates `result=nil, err=BifrostError` through the PostLLMHook. The governance plugin then evaluates `success = (nil != nil) = false` and skips billing.
- **Impact**: All token costs already processed by Anthropic before the stream was interrupted are lost.

### Scenario 2: Context cancelled before stream starts

- **Location**: `core/bifrost.go:5271`
- **Trigger**: Request is cancelled before the worker starts processing it
- **Mechanism**:
  ```go
  case <-ctx.Done():
      return nil, newBifrostCtxDoneError(ctx, "while waiting for stream response")
  ```
  Returns the error directly and **completely skips `RunPostLLMHooks()`**. The governance plugin is unaware that the request ever happened.
- **Impact**: Completely lost. The request may already have been queued and processed by the provider.

### Scenario 3: Non-streaming context cancellation

- **Location**: `core/bifrost.go:4973`
- **Trigger**: Client cancels while a non-streaming request is waiting for the provider's response
- **Mechanism**:
  ```go
  case <-ctx.Done():
      bifrostErr := newBifrostCtxDoneError(ctx, "waiting for provider response")
      return nil, bifrostErr
  ```
  Same as above — `RunPostLLMHooks()` is skipped, completely lost.
- **Impact**: Completely lost.

### Scenario 4: Request timeout (504)

- **Location**: `plugins/governance/tracker.go:72`
- **Trigger**: Request execution exceeds the timeout limit (fasthttp default 60s)
- **Mechanism**: `MakeRequestWithContext()` returns a 504 error → `result=nil` → `success=false` → billing skipped.
- **Impact**: Tokens already processed by Anthropic are not billed. Extended-thinking requests are particularly prone to timing out.

### Scenario 5: Provider returns a 500 error

- **Location**: `plugins/governance/tracker.go:72`
- **Trigger**: Anthropic returns a 500/502/503 or other server-side error
- **Mechanism**: The error response is converted to a `BifrostError` → `result=nil` → `success=false` → billing skipped.
- **Impact**: Anthropic may have already processed some tokens before returning the error (especially during input token processing); these costs are lost.

### Scenario 6: Intermediate failure during retries

- **Location**: `core/bifrost.go:5288` (the `executeRequestWithRetries` loop)
- **Trigger**: Request is retried because of a rate limit or network error
- **Mechanism**: Each failure inside the retry loop only triggers the next retry; **`PostLLMHook` is not invoked**. Only the final result (success or final failure) goes through the hook pipeline.
- **Impact**: Tokens that Anthropic already processed during intermediate failed attempts are not billed.

### Scenario 7: Client cancellation (499)

- **Location**: `core/providers/utils/utils.go:119-121`
- **Trigger**: Client actively cancels the request
- **Mechanism**: A code comment openly admits:
  > *"This function does NOT truly cancel the underlying fasthttp network request if the context is done. The fasthttp client call will continue in its goroutine until it completes or times out based on its own settings."*

  Bifrost stops waiting and returns 499, but the HTTP request continues in the background. Anthropic finishes processing the request and bills for it, while Bifrost records nothing.
- **Impact**: The cost of the entire request is lost.

### Scenario 8: postHookWorker panic

- **Location**: `plugins/governance/main.go:1586`
- **Trigger**: A panic occurs in `postHookWorker` or `UpdateUsage` (e.g., a nil pointer dereference)
- **Mechanism**: There is no `recover()` in the asynchronous goroutine. A panic causes **the whole process to crash**, and any in-memory counter data not yet flushed to disk is lost.
- **Impact**: Process crash + in-memory data loss (counters are flushed every 10 seconds, so up to 10 seconds of data can be lost).

---

## Summary Table

| # | Scenario | Code location | PostLLMHook called? | success value | Billing result |
|---|----------|---------------|---------------------|---------------|----------------|
| 1 | Streaming interruption | `main.go:1588` | Yes | false | **Lost** |
| 2 | ctx cancelled before stream starts | `bifrost.go:5271` | **No** | N/A | **Lost** |
| 3 | Non-streaming ctx cancellation | `bifrost.go:4973` | **No** | N/A | **Lost** |
| 4 | Timeout 504 | `tracker.go:72` | Yes | false | **Lost** |
| 5 | Provider 500 | `tracker.go:72` | Yes | false | **Lost** |
| 6 | Intermediate retry failure | `bifrost.go:5288` | **No** | N/A | **Lost** |
| 7 | Client cancellation 499 | `utils.go:119` | Yes | false | **Lost** |
| 8 | postHookWorker panic | `main.go:1586` | Yes | N/A | **Process crash** |

---

## Why the Impact Is Particularly Large in the Claude Code Scenario

Claude Code's usage pattern amplifies the issues above:

1. **Long streaming sessions** — Opus 4.6 + extended thinking lead to long single-request durations (potentially tens of seconds to minutes), increasing the chance of timeouts and interruptions.
2. **Frequent cancellations** — Ctrl+C in the terminal, switching files, closing the IDE, or network jitter all trigger ctx cancellations.
3. **Interruptions are more expensive under prompt caching** — Even when a request fails, cache creation tokens have already been billed by Anthropic ($6.25/MTok); a single request can involve hundreds of thousands of tokens.
4. **Network relay adds instability** — Going through a relay (54.64.152.0) adds an extra network hop, increasing the probability of disconnects.
5. **High-frequency interactions** — Claude Code's agentic mode automatically issues many consecutive requests, so the cumulative probability of failures and retries increases.

---

## Plausibility Analysis of the $30 Difference

Assume that over 30 days:
- Total requests ~1,000 (Claude Code's high-frequency interaction)
- 5–10% of them fail/are interrupted/time out for various reasons (~50–100 requests)
- Each failed request has on average processed ~$0.30–0.60 worth of tokens (input token processing + partial output)
- Cumulative: 50 × $0.60 = $30, which matches the observed difference.

---

### Steps to reproduce

`plugins/governance/tracker.go:72`:
```go
func (t *UsageTracker) UpdateUsage(ctx context.Context, update *UsageUpdate) {
    if !update.Success {
        t.logger.Debug("Request was not successful, skipping usage update")
        return  // ← all costs are lost
    }
    // ...
}
```

### Expected behavior

Should count all the cost if request already forwarded to LLM provider.

### Actual behavior

Failed requests skip billing.

### Affected area(s)

Plugins

### Version

v1.5.8

### Severity

Critical (blocks releases or production)

### PR DESCRIPTION
**Pull request by init0-lux**
## Summary

Currently, when a streaming request is cancelled mid-flight or fails, any partial token usage accumulated up to that point is silently discarded. Teams using budget enforcement lose visibility into these consumed-but-never-billed tokens, and the governance plugin's `postHookWorker` has no path to recover usage from errors. This PR builds an accounting pipeline that threads partial usage through `BifrostError` → stream cancellation/timeout handlers → `postHookWorker`, so that partially consumed tokens are always tracked against budgets.

Along the way, it also fixes three unrelated production issues discovered during testing: Gemini schema unmarshalling failures, a Bedrock TLS nil-logger panic, and incorrect Bedrock stop-reason folding for custom tools.

### Partial usage tracking for failed/cancelled streams (main feature)
- **`BifrostError` carries usage** — Added `Usage *BifrostLLMUsage` to `BifrostError.ExtraFields` so error paths can carry partial token counts.
- **All 11 streaming providers updated** — `HandleStreamCancellation` and `HandleStreamTimeout` in `core/providers/utils/utils.go` now accept a `*BifrostLLMUsage` parameter and attach it to the generated error. Every provider that streams (OpenAI, Anthropic, Bedrock, Gemini, Azure, Cohere, Mistral, ElevenLabs, Vertex, Replicate, vLLM) now declares `usage` at the top of its stream goroutine and passes it to both cancellation and timeout handlers.
- **`postHookWorker` rewired** — `PostLLMHook` now passes the `*BifrostError` into `postHookWorker`. The worker extracts usage from either the response or the error (`err.ExtraFields.Usage`), resolving the model from `err.ExtraFields.ResolvedModelUsed` when available. Cost is calculated via `CalculateCostFromUsage` regardless of success/failure.
- **`UsageTracker` early-return bug fixed** — The tracker previously returned immediately for any `Success == false` request. Now it only skips requests that are both unsuccessful **and** have zero usage (`TokensUsed == 0 && Cost == 0`).
- **New `CalculateCostFromUsage`** — Extracted from the response-bound `CalculateCost` into a standalone method on `ModelCatalog` that accepts a `*BifrostLLMUsage` + `PricingMetadata` struct, supporting all request types (text, embedding, speech, transcription, image, video, OCR, containers). This decouples pricing from the full response object so cost can be computed from raw usage alone.

### Gemini schema unmarshal fix
- Exported `ConvertFunctionParametersToSchema` (formerly unexported `convertFunctionParametersToSchema`) in `core/providers/gemini/utils.go` so test suites can deserialize JSON Schemas through intermediate `ToolFunctionParameters` objects, bypassing Go's strict tag validation on Gemini's wire-format string constraints.

### Bedrock TLS nil-logger panic fix
- Added defensive nil-logger guards in `ConfigureProxy` and `ConfigureTLS` (`core/providers/utils/utils.go`). When unit/integration tests initialize providers with `nil` loggers, the engine now safely falls back to the thread-safe `noopLogger` instead of panicking during `insecure_skip_verify` logging checks.

### Bedrock stop-reason fix
- Scoped the single-turn tool-use resolution loop in `ToBedrockConverseResponse` to only match server-managed system tools (`type == "server_tool_use"`). This preserves `end_turn` folding for built-in Nova tools while keeping the correct `tool_use` stop reason for client-managed custom tools.

### Test coverage
- `core/providers/utils/stream_test.go` — `TestHandleStreamCancellation_WithUsage` validates that `BifrostError.ExtraFields.Usage` is populated when a stream is cancelled with accumulated usage.
- `plugins/governance/tracker_test.go` — `TestUsageTracker_UpdateUsage_FailedRequestNoUsage` validates that truly empty failures (no usage) are still skipped, while the existing `TestUsageTracker_UpdateUsage_FailedRequest` test was updated to assert that partial usage **is** tracked.

## Type of change

- [x] Bug fix
- [x] Feature
- [x] Refactor

## Affected areas

- [x] Core (Go)
- [x] Providers/Integrations
- [x] Plugins

## How to test

```sh
# Governance plugin unit tests (tracker, cost calculation)
make test-plugins

# Stream cancellation with usage propagation
cd core/providers/utils && go test -run TestHandleStreamCancellation_WithUsage -v

# All provider integration tests (streaming coverage)
make test-core

# Specific provider streaming
make test-core PROVIDER=openai PATTERN=Stream
make test-core PROVIDER=anthropic PATTERN=Stream
make test-core PROVIDER=gemini PATTERN=Stream
```
## Breaking changes

- [x] No

## Related issues

Closes #3357 

## Security considerations

- Usage data is carried through the same `BifrostError` and `BifrostResponse` paths already subject to existing logging and observability controls.
- Nil-logger fallback uses the existing thread-safe `noopLogger` — no new logging paths are introduced.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [x] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)

## Summary by CodeRabbit

* **New Features**
  * Added response compaction support across AI providers
  * Enhanced cost calculation from token usage data
  * Improved governance team configuration with external source IDs
  * Added model blacklist support for provider configurations

* **Bug Fixes**
  * Fixed failed request handling to properly track partial usage data
  * Improved upstream connection error reporting

* **Infrastructure**
  * Enhanced private network connectivity options for provider configurations
  * Refined token usage tracking across streaming operations

### CODE REVIEW

**Comment by akshaydeo** [comment]
hi @init0-lux can you rebase your PR with the new changes

**Comment by init0-lux** [comment]
@akshaydeo Done! Please review

File: core/providers/anthropic/anthropic.go [hunk 1]
```diff
@@ -1250,42 +1253,42 @@ func HandleAnthropicResponsesStream(
...
-				if usageToProcess.OutputTokens > usage.OutputTokens {
-					usage.OutputTokens = usageToProcess.OutputTokens
+				if usageToProcess.OutputTokens > responseUsage.OutputTokens {
+					responseUsage.OutputTokens = usageToProcess.OutputTokens
 				}
-				calculatedTotal := usage.InputTokens + usage.OutputTokens
+				calculatedTotal := responseUsage.InputTokens + responseUsage.OutputTokens
 				if calculatedTotal > usage.TotalTokens {
 					usage.TotalTokens = calculatedTotal
 				}
```
[the remark this answers is not in the document]
Comment (init0-lux) [reply core/providers/anthropic/anthropic.go:1269]: accepted

File: core/providers/bedrock/bedrock.go [hunk 2]
```diff
@@ -1561,12 +1563,13 @@ func (provider *BedrockProvider) ResponsesStream(ctx *schemas.BifrostContext, po
...
+		usage := &schemas.BifrostLLMUsage{}
 		defer providerUtils.EnsureStreamFinalizerCalled(ctx, postHookSpanFinalizer)
 		defer func() {
 			if ctx.Err() == context.Canceled {
-				providerUtils.HandleStreamCancellation(ctx, postHookRunner, responseChan, provider.logger, postHookSpanFinalizer, jsonData)
+				providerUtils.HandleStreamCancellation(ctx, postHookRunner, responseChan, provider.logger, postHookSpanFinalizer, jsonData, usage)
 			} else if ctx.Err() == context.DeadlineExceeded {
-				providerUtils.HandleStreamTimeout(ctx, postHookRunner, responseChan, provider.logger, postHookSpanFinalizer, jsonData)
+				providerUtils.HandleStreamTimeout(ctx, postHookRunner, responseChan, provider.logger, postHookSpanFinalizer, jsonData, usage)
 			}
```
[the remark this answers is not in the document]
Comment (init0-lux) [reply core/providers/bedrock/bedrock.go:1574]: Added code at the end of the streaming accumulator block to copy accumulated token details (InputTokens, OutputTokens, TotalTokens, and CachedRead/WriteTokens) from the internal responseUsage to the output-facing usage struct.

File: framework/modelcatalog/pricing.go [hunk 3]
```diff
@@ -31,6 +31,85 @@ type PricingEntry struct {
...
+	// Normalize request type
+	requestType := normalizeStreamRequestType(metadata.RequestType)
+	// Handle custom pricing models like containers
+	lookupModel, lookupResolved := metadata.OriginalModel, metadata.ResolvedModel
+	if metadata.ContainerIdentifier != "" {
+		lookupModel = metadata.ContainerIdentifier
+		lookupResolved = metadata.ContainerIdentifier
+	}
+	// Resolve the pricing configuration
+	pricing := mc.resolvePricing(string(metadata.Provider), lookupModel, lookupResolved, requestType, s)
```
[the remark this answers is not in the document]
Comment (init0-lux) [reply framework/modelcatalog/pricing.go]: accepted

**Comment by init0-lux** [comment]
@akshaydeo @danpiths @roroghost17 made the changes that coderabbit suggested

**Comment by init0-lux** [comment]
@akshaydeo Shall I resolve the merge conflicts? There aren't any breaking changes

Or did #3941 fix this

**Comment by akshaydeo** [comment]
#3941 is different - you can replace the conflicts

**Comment by init0-lux** [comment]
@akshaydeo Resolved conflicts. 

13 were changelogs, 10 were formatting/whitespace related, 4 files had merges on both streams, 1 test file.

**Comment by akshaydeo** [comment]
@init0-lux apologies but this PR has a secret leak from your side - can you fix - rotate your internal secret and update the PR please

**Review by init0-lux — COMMENTED** [review]
@AkshayDeo I've gone through all 391 files, didnt find any leaks. I didnt use any secret key whatsoever during this PR's dev, and am positive.

Could you verify, and correct me if I'm wrong?

**Comment by akshaydeo** [comment]
you dont need to go through all files - please do the rebase - and find changes in your file. We got this alert in our secret scanner - will confirm once you rebase the PR

**Comment by init0-lux** [comment]
@akshaydeo updated. please check

**Comment by akshaydeo** [comment]
Hi @init0-lux — thanks for the contribution! This PR is currently blocked because our CLA bot shows the Contributor License Agreement as not yet signed. Could you sign it here so we can move this forward: https://cla-assistant.io/maximhq/bifrost?pullRequest=3709

Let us know if you run into any issues signing.

**Comment by jigish** [comment]
hi! is this planned for merge/release? we're waiting on this one to be able to use bifrost in production. robust canceled request accounting would be great!

**Comment by init0-lux** [comment]
> hi! is this planned for merge/release? we're waiting on this one to be able to use bifrost in production. robust canceled request accounting would be great!

@jigish I believe it was fixed in some PR by the core team, but if the issue's still there I'm happy to work on this and integrate this PR into the current code. @akshaydeo

**Comment by jigish** [comment]
> > hi! is this planned for merge/release? we're waiting on this one to be able to use bifrost in production. robust canceled request accounting would be great!
> 
> @jigish I believe it was fixed in some PR by the core team, but if the issue's still there I'm happy to work on this and integrate this PR into the current code. @akshaydeo

@init0-lux @akshaydeo I don't think this is fixed?

Anthropic streaming response, cancel the response mid-output, bifrost reports $0 for the request on v2.2.0. Am I missing something?

**Comment by akshaydeo** [comment]
I don't think this is true - we have regression tests for it; cancelled Anthropic stream carries cost with it. 

This is already fixed 

@jigish please share any screenshot that shows this otw - team will look into it. Closing this PR for now

**Comment by jigish** [comment]
@akshaydeo I had my claude code explain a bit further what I was seeing. here is what it said:

Thanks — you're right that there's a mechanism, and I should be precise about what I'm claiming. `attachBilledUsageFromContext` has been in `HandleStreamCancellation` since v2.2.0 and cancelled streams do produce a log row with a non-zero cost. I'm not disputing that.

But the cost it carries is **input-only**. The accumulated output is not in it.

Here are two requests against the same v2.2.1 instance, same provider and key, logging enabled (`client.enable_logging: true` + sqlite `logs_store`), read back from `/api/logs`:

```
status       deltas delivered   prompt_tokens   completion_tokens   cost_usd
success                   268              20                 300   0.00456
cancelled                 806              35                   1   0.00012
```

The completed stream delivered 268 `content_block_delta` events and logged 300 completion tokens — correct. The cancelled stream delivered **806** delta events, three times as much generated output, and logged **1** completion token. Its `cost_breakdown` is `input_cost 0.000105 / output_cost 0.000015`, i.e. the prompt plus the single token `message_start` seeds. On Sonnet pricing, ~806 deltas is on the order of 6k output tokens (~$0.09) against $0.000015 recorded.

So the partial generation isn't being billed — only the prompt is. The pipeline itself is fine: the success row proves it records output correctly when the stream completes.

**Second, separate issue: the telemetry plugin doesn't see any of it.**

```
plugins/logging/main.go       BilledUsage x10
plugins/governance/main.go    BilledUsage x4
plugins/telemetry/main.go     BilledUsage x0
```

`plugins/telemetry/main.go` derives every token and cost value from the response object (`result.ChatResponse.Usage`, `result.ResponsesResponse.Usage`, …). A cancelled stream yields a `BifrostError`, not a response, so those are nil. After the cancelled request above, the only series present is:

```
bifrost_error_requests_total  1
```

No `bifrost_cost_total`, no `bifrost_input_tokens_total`, no `bifrost_output_tokens_total` — those families don't exist at all until something increments them, and nothing does. The error counter moving shows the hook fired; it just had no usage to record. Enabling the logging plugin does not change this, since the two plugins consume different things.

That second point is what makes this load-bearing for us: our deployment runs telemetry without the logging plugin and without governance budgets, so Prometheus is our only accounting surface and cancelled-stream cost is invisible by construction. Anyone scraping `bifrost_cost_total` as their spend figure is undercounting, with no counter that would reveal it.

Both look addressable independently:
- have the stream accumulator keep the running output count so `BilledUsage` reflects what was generated, not just the `message_start` seed;
- have the telemetry post-hook fall back to `BilledUsage` when the response is absent.

Happy to open a PR for either if that's a direction you'd take.

**Repro** — config with one Anthropic provider key, plus `client.enable_logging: true` and a sqlite `logs_store`:

```bash
docker run -d -p 7099:7099 -e ANTHROPIC_API_KEY=$ANTHROPIC_API_KEY \
  -v ./config-logging.json:/app/data/config.json:ro \
  maximhq/bifrost:v2.2.1 -port 7099

# cancel a long stream mid-flight
curl -sN -X POST localhost:7099/anthropic/v1/messages \
  -H 'content-type: application/json' -H 'anthropic-version: 2023-06-01' \
  -d '{"model":"claude-sonnet-4-5-20250929","max_tokens":4096,"stream":true,
       "messages":[{"role":"user","content":"Write a long continuous history of the printing press. No headings."}]}' \
  | head -c 60000 >/dev/null

sleep 10
curl -s 'localhost:7099/api/logs?limit=2' | jq '.logs[] | {status, token_usage, cost}'
curl -s localhost:7099/metrics | grep '^bifrost_'
```
